### PR TITLE
Fixes bugs for Lufthansa, Twitter, Rumble

### PR DESCRIPTION
**Pull request by GioSensation**
**Reviewer:** @shakyShane 
**Asana:** https://app.asana.com/0/1198964220583541/1206821130893761/f

## Description
Fixes bugs for Lufthansa, Twitter, Rumble.
- [Lufthansa.com password doesn't autofil](https://app.asana.com/0/0/1206799448127350) 
- [twitter.com settings shows generate instead of fill](https://app.asana.com/0/0/1206442422819536/f/)
- [rumble.com slowing down with autofill enabled](https://app.asana.com/0/0/1206565642505594)

## Steps to test
Automated tests added. Plus tested on the real sites.

### CODE REVIEW

File: integration-test/helpers/mocks.js [hunk 1]
```diff
@@ -21,7 +21,7 @@ export const constants = {
         'loginWithFormInModal': `${privacyTestPagesPrefix}/autoprompt/2-form-in-modal.html`,
         'signupWithFormInModal': `${localPagesPrefix}/signup-in-modal.html`,
         'loginCovered': `${localPagesPrefix}/login-covered.html`,
-        'loginMultistep': `${localPagesPrefix}/login-multistep.html`,
+        'loginMultistep': `${privacyTestPagesPrefix}/autoprompt/3-multistep-form.html`,
```
Comment (GioSensation) [inline integration-test/helpers/mocks.js:24]: Another one moved to the privacy test pages repo 🎉

File: integration-test/tests/login-form.ios.spec.js [hunk 2]
```diff
@@ -104,7 +105,10 @@ test.describe('Auto-fill a login form on iOS', () => {
                 })
                 await login.promptWasShown()
                 await login.assertUsernameFilled(personalAddress)
+                // Password has no icon because it's disabled …
+                await genericPage(page).passwordHasNoIcon()
                 await login.assertPasswordEmpty()
+                // … yet filling should work once it's enabled
```
Comment (GioSensation) [inline integration-test/tests/login-form.ios.spec.js:111]: This covers the Lufthansa case.

File: src/Form/FormAnalyzer.js [hunk 3]
```diff
@@ -34,8 +34,16 @@ class FormAnalyzer {
          */
         this.signals = []
 
+        // Analyse the input that was passed. This is pretty arbitrary, but historically it's been working nicely.
         this.evaluateElAttributes(input, 1, true)
-        form ? this.evaluateForm() : this.evaluatePage()
+
+        // If we have a meaningful container (a form), check that, otherwise check the whole page
+        if (form !== input) {
```
Comment (GioSensation) [inline src/Form/FormAnalyzer.js:41]: This is a minor fix that regressed at some point. Largely inconsequential, but we were supposed to check if the input is === the form, instead we were checking for form existence. Nothing changed in the test suite, so it wasn't doing any harm. It's an edge case anyway. I've also expanded the conditional for clarity.

File: src/Form/FormAnalyzer.js [hunk 4]
```diff
@@ -291,13 +300,18 @@ class FormAnalyzer {
         this.evaluateElAttributes(this.form)
 
         // Check form contents (noisy elements are skipped with the safeUniversalSelector)
-        this.form.querySelectorAll(this.matching.cssSelector('safeUniversalSelector')).forEach(el => {
+        const formElements = this.form.querySelectorAll(this.matching.cssSelector('safeUniversalSelector'))
+        for (let i = 0; i < formElements.length; i++) {
+            // Safety cutoff to avoid huge DOMs freezing the browser
+            if (i >= 200) break
```
Comment (GioSensation) [inline src/Form/FormAnalyzer.js:306]: Stop at the first 200 elements, so that huge forms don't cause performance issues. This was the Rumble issue.

File: src/Form/inputTypeConfig.js [hunk 5]
```diff
@@ -78,12 +78,13 @@ const canBeInteractedWith = (input) => !input.readOnly && !input.disabled
  * @returns {Promise<boolean>}
  */
 const canBeAutofilled = async (input, device) => {
-    if (!canBeInteractedWith(input)) return false
```
Comment (GioSensation) [inline src/Form/inputTypeConfig.js:81]: This fixes Lufthansa. Basically, we don't want to decide whether we can autofill an input based on whether it can be interacted with. We use interactivity to decide whether to show the icon, but that's done elsewhere.

File: src/Form/inputTypeConfig.js [hunk 6]
```diff
@@ -129,15 +130,11 @@ const inputTypeConfig = {
...
-                }
-            }
-
-            // if we are on a 'login' page, check if we have data to autofill the field
-            if (isLogin || isHybrid || variant === 'current') {
+            if (
+                (subtype === 'password' && variant === 'new') || // New passord field
+                (isLogin || isHybrid || variant === 'current') // Current password field
+            ) {
+                // Check feature flags and available input types
```
Comment (GioSensation) [inline src/Form/inputTypeConfig.js:137]: I've moved the feature flag check to `canBeAutofilled` to keep concerns separate. This allowed me to simplify the checks here. I was almost able to consolidate all these functions to work the same, but then ran out of time.

File: src/Form/matching-config/selectors-css.js [hunk 7]
```diff
@@ -2,9 +2,13 @@
...
 
-const formInputsSelector = [
+const formInputsSelectorWithoutSelect = [
     genericTextField,
-    '[autocomplete=username]',
+    '[autocomplete=username]'
+]
+
+const formInputsSelector = [
+    ...formInputsSelectorWithoutSelect,
```
Comment (GioSensation) [inline src/Form/matching-config/selectors-css.js:11]: Separated the one with `select` and the one without. This allows us to skip `select` in the entry point scan, so if a page or a form only has a `select` element we don't waste any resources scanning something we can't fill anyway.

File: src/Form/matching-config/selectors-css.js [hunk 8]
```diff
@@ -249,20 +253,23 @@ const addressCountryCode = [
...
 const birthdayMonth = `
+[autocomplete=bday-month i],
 [name=bday-month i],
 [name*=birthday_month i], [name*=birthday-month i],
 [name=date_of_birth_month i], [name=date-of-birth-month i],
 [name^=birthdate_m i], [name^=birthdate-m i],
 select[name="mm" i]`
 
 const birthdayYear = `
+[autocomplete=bday-year i],
```
Comment (GioSensation) [inline src/Form/matching-config/selectors-css.js:272]: Just realised that I didn't include these. Doesn't really matter at this stage, but was a 0 cost thing to add 🙂.

File: test-forms/wired_checkout.html [hunk 9]
```diff
@@ -20,7 +20,7 @@
 										.prechecked{background-color:#fff; text-align: center; font-size: 16px;}
                                     </style>
                                     <div id="isgiftholder" style="pointer-events: none;">
-									<div id="isGiftCheck" aria-labelledby="isGiftCheckText" data-isgift="false" role="button" tabindex="0"><img id="nextgen-gift-check" src="https://subscribe.wired.com/circulation/wired/images/ng-radio-check-new.png" alt=""></div>
+									<div id="isGiftCheck" aria-labelledby="isGiftCheckText" data-isgift="false" role="button" tabindex="0"></div>
```
Comment (GioSensation) [inline test-forms/wired_checkout.html:23]: One thing I realised while looking into this is that we didn't remove images or scripts from these forms. Ivy did remove them when collecting the forms with the scraper, but we didn't do it for the other forms. Probably a nice win to remove everything from all these examples so tests run faster. Something for another time.

**Review by shakyShane — APPROVED** [review]
LGTM!
